subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -82,86 +86,84 @@ public <V> CrossBuildInMemoryCache<Class<?>, V> newClassMap() {
     }
 
     private abstract static class AbstractCrossBuildInMemoryCache<K, V> implements CrossBuildInMemoryCache<K, V>, BuildSessionLifecycleListener {
-        private final Object lock = new Object();
-        private final Map<K, V> valuesForThisSession = new HashMap<>();
+        private final ConcurrentHashMap<K, Lazy<V>> valuesForThisSession = new ConcurrentHashMap<>();
 
         @Override
         public void beforeComplete() {
-            synchronized (lock) {
-                retainValuesFromCurrentSession(valuesForThisSession.values());
-                valuesForThisSession.clear();
-            }
+            retainValuesFromCurrentSession(valuesForThisSession.values().stream().map(Lazy::get));
+            valuesForThisSession.clear();
         }
 
         @Override
         public void clear() {
-            synchronized (lock) {
-                valuesForThisSession.clear();
-                discardRetainedValues();
-            }
+            valuesForThisSession.clear();
+            discardRetainedValues();
         }
 
-        protected abstract void retainValuesFromCurrentSession(Collection<V> values);
+        protected abstract void retainValuesFromCurrentSession(Stream<V> values);
 
         protected abstract void discardRetainedValues();
 
+        /**
+         * Must be thread-safe.
+         */
         protected abstract void retainValue(K key, V v);
 
+        /**
+         * Must be thread-safe.
+         */
         @Nullable
         protected abstract V maybeGetRetainedValue(K key);
 
         @Nullable
         @Override
         public V getIfPresent(K key) {
-            synchronized (lock) {
-                return getIfPresentWithoutLock(key);
-            }
+            return valuesForThisSession
+                .computeIfAbsent(key, k -> Lazy.unsafe().of(() -> maybeGetRetainedValue(k)))
+                .get();
         }
 
+        /**
+         * Factory must be thread-safe and must not rely on thread-local state as it might
+         * be executed from a different thread than the one that submitted it.
+         */
         @Override
         public V get(K key, Function<? super K, ? extends V> factory) {
-            synchronized (lock) {
-                V v = getIfPresentWithoutLock(key);
-                if (v != null) {
-                    return v;
-                }
-
-                // TODO - do not hold lock while computing value
-                v = factory.apply(key);
-
-                retainValue(key, v);
-
-                // Retain strong reference
-                valuesForThisSession.put(key, v);
-
-                return v;
-            }
+            return valuesForThisSession.computeIfAbsent(key, k -> {
+                V retained = maybeGetRetainedValue(k);
+                return retained != null
+                    ? Lazy.unsafe().of(() -> retained)
```

Comment:
Not sure I love the idea of having a Lazy.unsafe and calling `get` outside of the `computeIfAbsent` even if it is _technically_ a "safe" function that we are passing into `Lazy.unsafe`. 

I would vote to create new `Lazy.ofValue` that represents an "already realized" value, which is always thread-safe as it is immutable and always returns the same value

I don't think it makes any difference in terms of thread-safety as this is already safe, but it feels wrong to use `unsafe` this way.

Reply:
> I would vote to create new `Lazy.ofValue` that represents an "already realized" value, which is always thread-safe as it is immutable and always returns the same value

I think that would make things clearer, yes.